### DISCUSSION TITLE
LIA+EUF problem where `cvc5` returns `unknown`

### DISCUSSION THREAD

**Discussion by nicola-gigante** [cvc5/cvc5#9345] [Q&A] [answered]
Hello,

consider this problem:

```
(set-logic ALL)

(declare-const z Int)
(declare-fun p (Int Int) Bool)

(assert (forall ((x Int) (y Int)) (= (p x y) (= z y))))

(check-sat)
```

`cvc5` returns `unknown` or hangs, depending on the options. The output:

```
$ cvc5 --lang=smtlib2 test.smt
unknown
$ cvc5 --lang=smtlib2 --finite-model-find test.smt
unknown
$ cvc5 --lang=smtlib2 --mbqi test.smt
<hangs forever....>
```

Z3 replies `sat`. 

I've tried the `--mbqi` option because @ajreynol suggested me to use it for similar formulas.

The question is: is this intrinsically difficult to solve for `cvc5` or is there any option that I can try?

**Comment by ajreynol** [comment, marked as answer]
One option that solves this `--macros-quant --macros-quant-mode=all`. 

This treats quantified formulas that specify definitions as macros, i.e. `p` is replaced by `(lambda ((x Int) (y Int)) (= z y))` in the example above. In fact, this makes the problem trivial.

Without this option, the challenge is that otherwise cvc5 is required to synthesize a specific definition of `p`, where unfortunately the `mbqi` option will not stumble upon `(= z y)`.

Another more expensive option that solves this is `--sygus-inference` which recasts the input problem as a SyGuS problem, enumerates the space of definitions until it finds `z->0, p->(= 0 y)`.

**Reply by nicola-gigante** [reply]
Thanks Andrew,

do the `--macros-quant --macros-quant-mode=all` have any drawback in general or can I safely specify them without worries?

Best,
Nicola

**Comment by ajreynol** [comment]
There are a few minor drawbacks:
- It may make preprocessing a bit slower, which can be an issue for very large inputs.
- It may confuse trigger-based instantiation, e.g. if `(p x y)` would have otherwise be used as a pattern.

If these are not a concern, I would not expect any major issues using these options.

**Reply by nicola-gigante** [reply]
Thanks,

I've tried it and it works for the formula in the question but it seems to be brittle. If I expand the `iff` as a double implication I get `unknown` again.
```
(set-logic ALL)

(declare-const z Int)
(declare-fun p (Int Int) Bool)

(assert (forall ((x Int) (y Int)) 
    (and (or (not (p x y)) (= z y))
         (or (p x y) (not (= z y))))
))

(check-sat)
```

```
$ cvc5 --lang=smtlib2 --macros-quant --macros-quant-mode=all test.smt
unknown
```

This is actually more similar to the encoding generated by my solver. 

The `--sygus-inference` option does indeed work in this case as well.

**Comment by ajreynol** [comment]
Yes, unfortunately this is expected.  This is indeed a limitation of the model generation capabilities of SMT solvers, z3 may have other heuristics for choosing models that the above benchmark works for, but it will have its own limitations.  Models enumerated by SyGuS will be less brittle, but will not scale as well.

**Reply by nicola-gigante** [reply]
I see, thanks!
